src/crypto/evp.h

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2022, The Monero Project
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+/**
+ * \file This header provides wrappers around the OpenSSL EVP API, specifically in the area of memory safety
+ */
+
+#include <openssl/evp.h>
+
+namespace crypto::evp {
+	/**
+	 * \class DigestContext is a memory-safe manager for EVP_MD_CTX. No need to call EVP_MD_CTX_{new,free}
+	 */
+	class DigestContext {
+	public:
+		/**
+		 * \brief Calls EVP_MD_CTX_new on its internal handle
+		 */
+		DigestContext() noexcept {
+			this->m_handle = EVP_MD_CTX_new();
+		}
+
+		/**
+		 * \brief Since recreating EVP_MD_CTXs is non-trival, copy construction is not allowed
+		 */
+		DigestContext(const DigestContext& other) = delete;
+
+		/**
+		 *\brief Since recreating EVP_MD_CTXs is non-trival, copy assignment is not allowed
+		 */
+		DigestContext& operator=(const DigestContext& other) = delete;
```

Comment:
```suggestion
		/**
		 * \brief Since recreating EVP_MD_CTXs is non-trival, copies and moves are not allowed.
		 */
		DigestContext& operator=(DigestContext&&) = delete;
```

The so-called 'desdemova' trick to disable both copies and moves.